src/monstergenerator.cpp

```diff
@@ -659,30 +659,49 @@ void MonsterGenerator::load_monster( const JsonObject &jo, const std::string &sr
     mon_templates->load( jo, src );
 }
 
-mon_effect_data load_mon_effect_data( const JsonObject &e )
-{
-    return mon_effect_data( efftype_id( e.get_string( "id" ) ), e.get_int( "duration", 0 ),
-                            e.get_bool( "affect_hit_bp", false ),
-                            bodypart_str_id( e.get_string( "bp", "bp_null" ) ),
-                            e.get_bool( "permanent", false ),
-                            e.get_int( "chance", 100 ) );
-}
 
-class mon_attack_effect_reader : public generic_typed_reader<mon_attack_effect_reader>
+mon_effect_data::mon_effect_data() :
+    chance( 100.0f ),
+    permanent( false ),
+    affect_hit_bp( false ),
+    bp( body_part_bp_null ),
+    duration( 1, 1 ),
+    intensity( 0, 0 ) {}
+
+void mon_effect_data::load( const JsonObject &jo )
 {
-    public:
-        mon_effect_data get_next( JsonValue &jv ) const {
-            JsonObject e = jv.get_object();
-            return load_mon_effect_data( e );
-        }
-        template<typename C>
-        void erase_next( std::string &&eff_str, C &container ) const {
-            const efftype_id id = efftype_id( std::move( eff_str ) );
-            reader_detail::handler<C>().erase_if( container, [&id]( const mon_effect_data & e ) {
-                return e.id == id;
-            } );
-        }
-};
+    assign( jo, "id", id );
+
+    if( jo.has_float( "chance" ) ) {
+        assign( jo, "chance", chance, false, 0.0f, 100.0f );
+    }
+    if( jo.has_bool( "permanent" ) ) {
+        assign( jo, "permanent", permanent );
+    }
+    if( jo.has_bool( "affect_hit_bp" ) ) {
+        assign( jo, "affect_hit_bp", affect_hit_bp );
+    }
+    if( jo.has_string( "bp" ) ) {
+        assign( jo, "bp", bp );
+    }
+    if( jo.has_string( "message" ) ) {
+        assign( jo, "message", message );
+    }
+
+    // Support shorthand for a single value.
+    if( jo.has_int( "duration" ) ) {
+        int i = jo.get_int( "duration", 0 );
+        duration = { i, i };
+    } else if( jo.has_array( "duration" ) ) {
+        assign( jo, "duration", duration );
+    }
+    if( jo.has_int( "intensity" ) ) {
+        int i = jo.get_int( "intensity", 0 );
+        intensity = { i, i };
+    } else if( jo.has_array( "intensity" ) ) {
+        assign( jo, "intensity", intensity );
+    }
```

Comment:
```suggestion
    mandatory( jo, false, "id", id );
    optional( jo, false, "chance", chance, 100.f );
    optional( jo, false, "permanent", permanent, false );
    optional( jo, false, "affect_hit_bp", affect_hit_bp, false );
    optional( jo, false, "bp", bp, body_part_bp_null );
    optional( jo, false, "message", message );
    // Support shorthand for a single value.
    if( jo.has_int( "duration" ) ) {
        int i = 1;
        mandatory( jo, false, "duration", i );
        duration = { i, i };
    } else {
        optional( jo, false, "duration", duration, std::pair<int, int> { 1, 1 } );
    }
    if( jo.has_int( "intensity" ) ) {
        int i = 1;
        mandatory( jo, false, "intensity", i );
        intensity = { i, i };
    } else {
        optional( jo, false, "intensity", intensity, std::pair<int, int> { 1, 1 } );
    }

    if( chance > 100.f || chance < 0.f ) {
        jo.throw_error_at( "chance",
                           string_format( "\"chance\" is defined as %f, "
                                          "but must be a decimal number between 0.0 and 100.0", chance ) );
        chance = clamp<float>( chance, 0.f, 100.f );
    }
```
`assign` is on the chopping block, so we'd be in a better spot using generic_factory methods like `optional` and `mandatory` moving forward